docs/proposals/push-metrics.md

```diff
@@ -0,0 +1,221 @@
+/api/v1/model/namespaces/webapp/metrics/custom/http_gatherer/frontend/hits_total
+/api/v1/model/namespaces/webapp/metrics/custom/http_gatherer/restapi/hits_total
+```
+
+## Discussed Alternatives
+
+A number of alternatives came up during the discussion of this proposal.  They
+are discussed briefly below.  Note that most of these alternatives do not deal
+particularly well with a case where metrics need to come from a source that is
+not running as a pod on the cluster.  While it is expected that many of the
+producers will be running as components on the cluster (e.g. as DaemonSets or
+PetSets), it could still be adventageous to support metrics coming from
+components that are not in the form of pods.
+
+### Writing directly into sinks
+
+This alternative would have producers write directly into sinks in the Heapster
+storage schema, and then use a mechanism similar to the Oldtimer API to read
+the metrics back.
+
+This would require every producer to know how to talk to every sink, would make
+configuring the sinks more complicated, and would most likely lead to software
+only being able to talk to one of the sinks supported by Heapster.
+Additionally, you lose the benefits of the Heapster model, and either have to
+adapt the existing Heapster model to fall back to an Oldtimer-like approach, or
+teach all cluster components to be able to use both the Heapster model and
+Oldtimer simultaneously.
+
+### Reworking the existing cAdvisor-Kubelet-Heapster Pull Mechanism
+
+This alternative would involve reworking the existing pull mechanism to allow
+certain pods to produce the "cluster-scoped" metrics (i.e. metrics scope to
+pods other than the producer), as opposed to the current situation, where all
+custom metrics collected via the current pull mechanism are scoped to the
+producer pod.
+
+This would require a mechanism for indicating to Heapster which pod names were
+allowed to produce cluster-scoped metrics, since admins would generally want
+most pods producing metrics to continue to just have producer-pod-scoped
+metrics.  It would also conceptually blend together pods producing metrics
+about themselves versus pods producing metrics about others.  Additionally, the
+current cAdvisor-based custom metrics collection is not secured, so all metrics
+would be available to anyone with knowlege of the appropriate port, but this
+may change in the future.
+
+### Using a new daemon per node to produce metrics
+
+This alternative would involve running a daemon on each node that aggregated
+all the separate custom metrics producers' results together.
+
+It was suggested that an approach similar to the Prometheus Node Exporter
+Textfile Collector could be used, in which sources would write their metrics to
+files in a directly, which would later be read by the collector when polled for
+metrics.  Using a shared directory like this poses issues when you want to run
+the producers in containers.  Instead, you could instead have producers push to
```

Comment:
You could do it in a number of ways depending on what you want. You could have one per pod, one per node or one per service.

Service ones are actually a bit messy with this idea, though for autoscaling I'd be looking for pod rather than service-level metrics. Service-level metrics imply that aggregation has occurred, which means there's an extra hop in the control loop - which you want to avoid.